### PR TITLE
Added TF device

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Adds support for the Strawberry Fields `tf` backend via PennyLane

**Benefits:**

* The new `strawberryfields.tf` device can be used with PennyLane's `PassthruQNode`, and thus supports `diff_method="backprop"`.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by dnnagy** [comment]
Hi all!

I would like to simulate a CV quantum circuit on the gpu and calculate gradients with this plugin. The official releases as far as I know do not include the TF backend feature, but I see that You are already working on it. Can I somehow use the package in this phase, without waiting for the release?

I have PennyLane 0.10.0 and PennyLane-SF 0.9.0 installed.

Thanks, in advance

File: doc/conf.py [hunk 1]
```diff
@@ -22,44 +22,6 @@
...
-               "log": mock_math_fn,
-               "matmul": mock_math_fn,
-               "Tensor": list,
-               "Variable": list,
-               "ndarray": list}
-
-mock = Mock(**mock_fns)
-for mod_name in MOCK_MODULES:
-    sys.modules[mod_name] = mock
-
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/conf.py:62]: Ah! previously, the pl-sf plugin didn't use TF as a requirement, so we simply mocked out TensorFlow in the documentation. However, this is no longer the case; we now _need_ to import TensorFlow to build the documentation, so this has to be removed.

File: doc/index.rst [hunk 2]
```diff
@@ -43,6 +36,16 @@ PennyLane-SF provides two Strawberry Fields devices for PennyLane:
     <div style='clear:both'></div>
     </br>
 
+.. devicegalleryitem::
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/index.rst:39]: It is deliberate, unfortunately it looked really ugly and squished having it on a single row 👍 

> Also, maybe update PennyLane-SF provides two Strawberry Fields devices for PennyLane: to four?


Nice, I missed that!

File: pennylane_sf/tf.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+import numpy as np
+import tensorflow as tf
+
+import strawberryfields as sf
+from strawberryfields.backends.tfbackend.states import FockStateTF
+
+# import state preparations
+from strawberryfields.ops import (
+    # Catstate,
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:29]: No, this is something we overlooked when we updated SF to only allow real parameters to all gates; the `Catstate` gate still accepts a complex argument! So unfortunately I couldn't get it to work with the TF2 backend (@antalszava)

File: pennylane_sf/tf.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    if N == len(device_wires):
+        # expectation value of the entire system
+        ex = state.fock_prob(n)
+        return ex, ex - ex ** 2
+
+    dm = state.reduced_dm(modes=device_wires.tolist())
+    ex = tf.math.real(dm[tuple([n[i // 2] for i in range(len(n) * 2)])])
+
+    var = ex - ex ** 2
+    return ex, var
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:119]: > Is it because the state isn't a BaseFockState, and thus fock_state doesn't return the correct values? 

No, actually the `FockStateTF` class that the `tf` backend returns is a subclass of `BaseFockState`, so `isinstance(state, BaseFockState)` would still work!

The issue is solely the `.real` in the other function; we need to replace all NumPy operations with TensorFlow operations (`tf.math.real`) to preserve differentiability, otherwise the gradient will just be `None`.

File: pennylane_sf/tf.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            # store parameters
+            param_labels = [str(uuid.uuid4()) for _ in range(len(par))]
+
+            for l, v in zip(param_labels, par):
+                self.params[l] = v
+
+            par = self.prog.params(*param_labels)
+
+            if not isinstance(par, Sequence):
+                par = (par,)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:224]: This is because, in the next line, I am doing `*par`, which will fail if `par` is not a sequence 😆 

I could have alternatively done
```python
if isinstance(par, Sequence):
    op = self._operation_map[operation](*par)
else:
    op = self._operation_map[operation](par)
```
but I felt simply always making `par` a sequence was more elegant?

File: pennylane_sf/tf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisac) [reply pennylane_sf/tf.py:119]: Oh, that makes more sense. I missed that change. Thanks! 😄

File: pennylane_sf/tf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thisac) [reply pennylane_sf/tf.py:224]: Oh, I meant the whole section (from line 214). 😊

**Comment by josh146** [comment]
Hi @dnnagy! Since this is still a work in progress, it has not made today's release. However, it is quite well tested, so should be pretty stable/usable! To use it, you will need the latest release of PennyLane (v0.11), the development version of Strawberry Fields (from GitHub), and this PR of PennyLane-SF.

This should be doable over pip:

```console
pip install pennylane==0.11
pip install git+https://github.com/XanaduAI/strawberryfields.git
pip install git+https://github.com/PennyLaneAI/pennylane-sf.git@tf2
```

File: pennylane_sf/tf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:224]: Oh! Right. Here I am automating the creation of the Strawberry Fields symbolic parameters, which are needed for the TF2 backend to correctly accept tf.Tensor objects:

image

To name them, I am taking a rather brute-force approach and just using `uuid` to generate hashes (and hoping there are no clashes). Is there a better way of doing this?

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:224]: The `self.prog.params` dictionary simply keeps track of the `{param_label: param_value}` so we can pass them on engine run.

[the remark this answers is not in the document]
Comment (thisac) [reply pennylane_sf/tf.py:224]: I see. Thanks for the explanation! :slightly_smiling_face:

File: pennylane_sf/tf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane_sf/tf.py:29]: Hmm, I'd imagine that `Catstate` was skipped because it didn't necessarily require a change in Strawberry Fields, right? :thinking: (That is no casting was needed there as calculations were done using complex values).

There seems to be a difference between the signatures of `Catstate` in PennyLane and SF: [PL turned to use real arguments due to parameter domain checks](https://github.com/PennyLaneAI/pennylane/pull/148) whereas SF remained with using complex arguments.

Think there are two alternatives:
1. Making up for the different signatures in this plugin, similarly to how is being done in [`simulator.py`](https://github.com/PennyLaneAI/pennylane-sf/blob/c152793b04c1a90f55e29440054828b872b99a6e/pennylane_sf/simulator.py#L103). The Fock device can currently handle `CatState` this way.

This solution includes adding logic to `StrawberryFieldsTF.apply` (worked locally for `expval`, for `probs` seems to take a bit more tweaking with the subsequent reshaping):
```python
    if operation == "CatState":
        par = (par[0] * np.exp(par[1] * 1j), par[2])
```
2. [Changing the signature in Strawberry Fields](https://github.com/XanaduAI/strawberryfields/pull/441).

File: tests/test_tf.py [hunk 6]
```diff
@@ -0,0 +1,1050 @@
...
+            qml.Displacement(x, 0, wires=0)
+            return qml.expval(qml.NumberOperator(0))
+
+        x = 1
+
+        runs = []
+        for _ in range(100):
+            runs.append(circuit(x))
+
+        expected_var = np.sqrt(1 / shots)
```
Comment (trbromley) [inline tests/test_tf.py:137]:
```suggestion
        expected_var = np.sqrt(1 / (shots * 100))
```
I'm not sure though

**Comment by trbromley** [comment]
Also, worth adding to the changelog?

File: doc/devices/tf.rst [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+The device can then be used just like other devices for the definition and evaluation of QNodes
+within PennyLane.
+
+For instance, the following simple example defines a QNode that first displaces the vacuum state,
+applies a beamsplitter, and then returns the marginal probability on the first wire. This function
+is then converted into a QNode which is placed on the :code:`strawberryfields.tf` device:
+
+.. code-block:: python
+
+    @qml.qnode(dev, interface="tf")
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/devices/tf.rst:43]: You need to use `interface="tf"` 🙂 

If you don't specify the interface, the default is autograd, which means the parameter-shift rule will be used, with autograd arrays converted to standard numpy arrays at the interface.

[the remark this answers is not in the document]
Comment (josh146) [reply doc/devices/tf.rst:43]: We _could_ also define the `StrawberryFieldsTF.jacobian()` method, which would simply execute the device within a gradient tape, and return `tape.jacobian(result, weights)`. This way:

* `interface="tf"`: the entire computation is a single backpropagation.

* `interface="autograd"` or `interface="torch"`: Outside the QNode, torch/autograd backprop is performed. When the backprop reaches the QNode, the custom gradient of the QNode is called.... which simply does an internal tensorflow backprop to get the jacobian matrix.

This is how `default.tensor.tf` acts.

File: doc/devices/tf.rst [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    the Fock basis truncation when applying quantum operations
+
+``hbar=2``
+    The convention chosen in the canonical commutation relation :math:`[x, p] = i \hbar`.
+    Default value is :math:`\hbar=2`.
+
+``analytic=True``
+    Indicates if the device should calculate expectations and variances analytically.
+    Note that backpropagation is not supported when ``analytic=False``; returned gradients
+    and Jacobians will be ``None``.
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/devices/tf.rst:128]: That's a good question, I went back and forth myself here.

The reason I ended up keeping it in: I can imagine a case where a user trains using this device analytically, and then wants to evaluate their trained model non-analytically without having to modify their code too much.

Although, we could remove this (_forcing_ the device to always be `analytic=True`). If the user then wants to evaluate their trained model non-analytically, they would be forced to change the device.

File: doc/index.rst [hunk 9]
```diff
@@ -21,7 +14,7 @@ straight away in PennyLane, without the need to import any additional packages.
 Devices
 =======
 
-PennyLane-SF provides two Strawberry Fields devices for PennyLane:
+PennyLane-SF provides four Strawberry Fields devices for PennyLane:
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/index.rst:17]:
```suggestion
PennyLane-SF provides various Strawberry Fields devices for PennyLane:
```

File: pennylane_sf/tf.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:1]: Nope, as TF is not a strict dependency of this device. Forcing this device to be imported on plugin import would cause issues for users that don't have TF installed.

File: pennylane_sf/tf.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    Args:
+        wires (int, Iterable[Number, str]]): Number of subsystems accessible on the device,
+            or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
+            or strings (``['ancilla', 'q1', 'q2']``).
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+        cutoff_dim (int): Fock-space truncation dimension
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:136]:
```suggestion
        shots (int): Number of circuit evaluations/random samples used
            to estimate expectation values of observables. If ``analytic=True``,
            this setting is ignored when calculating expectation values.
```

File: pennylane_sf/tf.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+            probs = tf.reshape(self.state.all_fock_probs(cutoff=cutoff), -1)
+
+        else:
+            rdm = self.state.reduced_dm(modes=device_wires.tolist())
+            new_state = FockStateTF(rdm, N, pure=False, cutoff_dim=cutoff)
+            probs = tf.reshape(new_state.all_fock_probs(cutoff=cutoff), -1)
+
+        ind = np.indices([cutoff] * N).reshape(N, -1).T
+        probs = OrderedDict((tuple(k), v) for k, v in zip(ind, probs))
+        return probs
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:275]: yep!

File: tests/test_tf.py [hunk 13]
```diff
@@ -0,0 +1,1050 @@
...
+        assert isinstance(circuit, qml.qnodes.PassthruQNode)
+
+        res = circuit(1)
+        assert isinstance(res, tf.Tensor)
+        assert np.allclose(res, 1, atol=tol, rtol=0)
+
+    def test_nonzero_shots(self):
+        """Test that the tf plugin provides correct result for high shot number"""
+        shots = 10 ** 2
+        dev = qml.device("strawberryfields.tf", wires=1, cutoff_dim=10, shots=shots)
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_tf.py:124]: good catch!
```suggestion
        dev = qml.device("strawberryfields.tf", wires=1, cutoff_dim=10, shots=shots, analytic=False)
```

File: doc/devices/tf.rst [hunk 14]
```diff
@@ -0,0 +1,194 @@
+The TF device
+=============
+
+The ``strawberryfields.tf`` device gives access to Strawberry Field's ``tf`` state simulator
+backend. This simulator device has the following features:
+
+* The simulator is written using TensorFlow, so supports classical backpropagation using
+  PennyLane. Simply use ``interface="tf"`` when creating your QNode; ``diff_method="backprop"``
+  will be set by default.
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/devices/tf.rst:9]:
```suggestion
  PennyLane. Simply use ``interface="tf"`` when creating your QNode.
```

File: pennylane_sf/tf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:29]: Update: after looking into this, it seems that only (2) will work

File: doc/devices/tf.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/devices/tf.rst:42]: @trbromley I've now defined the `jacobian()` method :)

File: pennylane_sf/tf.py [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+                new_queue.append(new_op)
+
+            self.reset()
+
+            res = self.execute(new_queue, observables, parameters=parameters)
+            res = tf.cast(tf.squeeze(tf.stack(res)), dtype=tf.float64)
+
+        jac = tape.jacobian(res, variables, experimental_use_pfor=False)
+        jac = tf.stack([i if i is not None else tf.zeros(res.shape, dtype=tf.float64) for i in jac])
+        return jac.numpy().T
```
Comment (trbromley) [inline pennylane_sf/tf.py:352]: That was a tough one! Thanks for the comments!

File: requirements.txt [hunk 16]
```diff
@@ -1,2 +1,3 @@
 strawberryfields>=0.15
 pennylane>=0.11
+tensorflow
```
Comment (trbromley) [inline requirements.txt:3]: Do we need to version pin to at least 2?
Also, the idea to not put in `setup.py` is it's a soft requirement?

File: pennylane_sf/tf.py [hunk 17]
```diff
@@ -273,3 +272,81 @@ def probability(self, wires=None):
...
+                # free parameters will be represented as Variable instances.
+                op_params[operation] = operation.data[:]
+
+            # Loop through the free parameter reference dictionary
+            for _, par_dep_list in parameters.items():
+                if not par_dep_list:
+                    # parameter is not used within circuit
+                    v = tf.Variable(0, dtype=tf.float64)
+                    variables.append(v)
+                    continue
```
Comment (thisac) [inline pennylane_sf/tf.py:296]: Nice! I haven't really used `continue` much, but it seems quite handy.

File: pennylane_sf/tf.py [hunk 18]
```diff
@@ -273,3 +272,81 @@ def probability(self, wires=None):
         ind = np.indices([cutoff] * N).reshape(N, -1).T
         probs = OrderedDict((tuple(k), v) for k, v in zip(ind, probs))
         return probs
+
+    def jacobian(
+        self, queue, observables, parameters
+    ):  # pylint: disable=missing-function-docstring
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_sf/tf.py:278]: Yep, unfortunately this is black's doing :(

File: pennylane_sf/tf.py (hunk 17 above)
Comment (josh146) [reply pennylane_sf/tf.py:295]: Yeah! Typically any pattern of the form

```python
for i in sequence:
    if condition:
        do_thing()
    else:
        do_other_thing()
```

can be re-written as

```python
for i in sequence:
    if condition:
        do_thing()
        continue
    
    do_other_thing()
```

Comes down to personal preference :) If the `else` is particular large, I tend to favour the latter just to avoid the additional indentation.

File: requirements.txt (hunk 16 above)
Comment (josh146) [reply requirements.txt:3]:
```suggestion
tensorflow>=2.0
```
>  Also, the idea to not put in setup.py is it's a soft requirement?
Yep, exactly :)

File: doc/devices/tf.rst [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+    :nosignatures:
+
+    ~pennylane.Identity
+    ~pennylane.NumberOperator
+    ~pennylane.TensorN
+    ~pennylane.X
+    ~pennylane.P
+    ~pennylane.QuadOperator
+    ~pennylane.PolyXP
+    ~pennylane.TensorN
```
Comment (antalszava) [inline doc/devices/tf.rst:189]:
```suggestion

```

File: doc/requirements.txt [hunk 20]
```diff
@@ -3,4 +3,5 @@ ipykernel
 nbsphinx
 pygments-github-lexers
 sphinx-automodapi
+tensorflow
```
Comment (antalszava) [inline doc/requirements.txt:6]: In PL we pin to `tensorflow==2.0.1`, would that be needed here too? :thinking:

Comment (josh146) [reply doc/requirements.txt:6]:
```suggestion
tensorflow>=2.0.1
```
